File: SPI/src/12/bit_banging_spi_implementation.c
// Bit Banging SPI Implementation in C
// This example shows a complete software SPI implementation

#include <stdint.h>
#include <stdbool.h>

// Pin definitions (adjust for your hardware)
#define SPI_MOSI_PIN    17
#define SPI_MISO_PIN    16
#define SPI_SCK_PIN     18
#define SPI_CS_PIN      5

// SPI Mode definitions
typedef enum {
    SPI_MODE0 = 0,  // CPOL=0, CPHA=0
    SPI_MODE1 = 1,  // CPOL=0, CPHA=1
    SPI_MODE2 = 2,  // CPOL=1, CPHA=0
    SPI_MODE3 = 3   // CPOL=1, CPHA=1
} spi_mode_t;

// Configuration structure
typedef struct {
    uint8_t mosi_pin;
    uint8_t miso_pin;
    uint8_t sck_pin;
    uint8_t cs_pin;
    spi_mode_t mode;
    uint32_t delay_us;  // Delay for clock timing
} bitbang_spi_t;

// Hardware abstraction layer functions (implement for your platform)
static inline void gpio_set_high(uint8_t pin) {
    // Platform-specific: Set GPIO pin high
    // Example: GPIO_REG |= (1 << pin);
}

static inline void gpio_set_low(uint8_t pin) {
    // Platform-specific: Set GPIO pin low
    // Example: GPIO_REG &= ~(1 << pin);
}

static inline bool gpio_read(uint8_t pin) {
    // Platform-specific: Read GPIO pin
    // Example: return (GPIO_REG & (1 << pin)) != 0;
    return false;
}

static inline void delay_microseconds(uint32_t us) {
    // Platform-specific: Delay in microseconds
    // Example: for (volatile uint32_t i = 0; i < us * CPU_MHZ; i++);
}

// Initialize bit banging SPI
void bitbang_spi_init(bitbang_spi_t *spi) {
    // Configure pins as outputs/inputs
    // MOSI, SCK, CS as outputs
    // MISO as input
    
    // Set initial states
    gpio_set_low(spi->mosi_pin);
    gpio_set_high(spi->cs_pin);  // CS is active low
    
    // Set clock idle state based on mode
    if (spi->mode == SPI_MODE2 || spi->mode == SPI_MODE3) {
        gpio_set_high(spi->sck_pin);  // CPOL=1
    } else {
        gpio_set_low(spi->sck_pin);   // CPOL=0
    }
}

// Transfer a single byte (full duplex)
uint8_t bitbang_spi_transfer_byte(bitbang_spi_t *spi, uint8_t data_out) {
    uint8_t data_in = 0;
    bool cpol = (spi->mode == SPI_MODE2 || spi->mode == SPI_MODE3);
    bool cpha = (spi->mode == SPI_MODE1 || spi->mode == SPI_MODE3);
    
    for (int i = 7; i >= 0; i--) {  // MSB first
        // Set MOSI based on data bit
        if (data_out & (1 << i)) {
            gpio_set_high(spi->mosi_pin);
        } else {
            gpio_set_low(spi->mosi_pin);
        }
        
        if (cpha == 0) {
            // CPHA=0: Sample on first edge, setup on second
            delay_microseconds(spi->delay_us);
            
            // First edge (sample)
            if (cpol) {
                gpio_set_low(spi->sck_pin);
            } else {
                gpio_set_high(spi->sck_pin);
            }
            
            // Read MISO
            if (gpio_read(spi->miso_pin)) {
                data_in |= (1 << i);
            }
            
            delay_microseconds(spi->delay_us);
            
            // Second edge (setup)
            if (cpol) {
                gpio_set_high(spi->sck_pin);
            } else {
                gpio_set_low(spi->sck_pin);
            }
        } else {
            // CPHA=1: Setup on first edge, sample on second
            delay_microseconds(spi->delay_us);
            
            // First edge (setup)
            if (cpol) {
                gpio_set_low(spi->sck_pin);
            } else {
                gpio_set_high(spi->sck_pin);
            }
            
            delay_microseconds(spi->delay_us);
            
            // Second edge (sample)
            if (cpol) {
                gpio_set_high(spi->sck_pin);
            } else {
                gpio_set_low(spi->sck_pin);
            }
            
            // Read MISO
            if (gpio_read(spi->miso_pin)) {
                data_in |= (1 << i);
            }
        }
    }
    
    return data_in;
}

// Transfer multiple bytes
void bitbang_spi_transfer(bitbang_spi_t *spi, 
                          const uint8_t *tx_data, 
                          uint8_t *rx_data, 
                          size_t length) {
    // Assert chip select (active low)
    gpio_set_low(spi->cs_pin);
    delay_microseconds(1);
    
    // Transfer each byte
    for (size_t i = 0; i < length; i++) {
        uint8_t tx_byte = tx_data ? tx_data[i] : 0xFF;
        uint8_t rx_byte = bitbang_spi_transfer_byte(spi, tx_byte);
        
        if (rx_data) {
            rx_data[i] = rx_byte;
        }
    }
    
    // Deassert chip select
    delay_microseconds(1);
    gpio_set_high(spi->cs_pin);
}

// Example usage: Read from SPI device
void example_read_register(void) {
    bitbang_spi_t spi = {
        .mosi_pin = SPI_MOSI_PIN,
        .miso_pin = SPI_MISO_PIN,
        .sck_pin = SPI_SCK_PIN,
        .cs_pin = SPI_CS_PIN,
        .mode = SPI_MODE0,
        .delay_us = 1  // 500 kHz effective speed
    };
    
    bitbang_spi_init(&spi);
    
    // Read register 0x0F
    uint8_t tx_buffer[2] = {0x0F | 0x80, 0x00};  // Read command + dummy byte
    uint8_t rx_buffer[2];
    
    bitbang_spi_transfer(&spi, tx_buffer, rx_buffer, 2);
    
    uint8_t register_value = rx_buffer[1];
    // Process register_value...
}

// Example usage: Write to SPI device
void example_write_register(uint8_t reg_addr, uint8_t value) {
    bitbang_spi_t spi = {
        .mosi_pin = SPI_MOSI_PIN,
        .miso_pin = SPI_MISO_PIN,
        .sck_pin = SPI_SCK_PIN,
        .cs_pin = SPI_CS_PIN,
        .mode = SPI_MODE0,
        .delay_us = 1
    };
    
    bitbang_spi_init(&spi);
    
    uint8_t tx_buffer[2] = {reg_addr & 0x7F, value};  // Write command + data
    
    bitbang_spi_transfer(&spi, tx_buffer, NULL, 2);
}

File: SPI/src/12/bit_banging_spi_implementation.rs
// Bit Banging SPI Implementation in Rust
// Safe, type-safe implementation with embedded-hal traits

use std::thread;
use std::time::Duration;

// GPIO Pin trait (abstract hardware interface)
pub trait OutputPin {
    type Error;
    fn set_high(&mut self) -> Result<(), Self::Error>;
    fn set_low(&mut self) -> Result<(), Self::Error>;
}

pub trait InputPin {
    type Error;
    fn is_high(&self) -> Result<bool, Self::Error>;
    fn is_low(&self) -> Result<bool, Self::Error>;
}

// SPI Mode enumeration
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SpiMode {
    Mode0,  // CPOL=0, CPHA=0
    Mode1,  // CPOL=0, CPHA=1
    Mode2,  // CPOL=1, CPHA=0
    Mode3,  // CPOL=1, CPHA=1
}

impl SpiMode {
    fn cpol(&self) -> bool {
        matches!(self, SpiMode::Mode2 | SpiMode::Mode3)
    }
    
    fn cpha(&self) -> bool {
        matches!(self, SpiMode::Mode1 | SpiMode::Mode3)
    }
}

// SPI Configuration
#[derive(Debug, Clone)]
pub struct SpiConfig {
    pub mode: SpiMode,
    pub delay_us: u64,
    pub msb_first: bool,
}

impl Default for SpiConfig {
    fn default() -> Self {
        Self {
            mode: SpiMode::Mode0,
            delay_us: 1,
            msb_first: true,
        }
    }
}

// Bit Banging SPI structure
pub struct BitBangSpi<MOSI, MISO, SCK, CS>
where
    MOSI: OutputPin,
    MISO: InputPin,
    SCK: OutputPin,
    CS: OutputPin,
{
    mosi: MOSI,
    miso: MISO,
    sck: SCK,
    cs: CS,
    config: SpiConfig,
}

impl<MOSI, MISO, SCK, CS> BitBangSpi<MOSI, MISO, SCK, CS>
where
    MOSI: OutputPin,
    MISO: InputPin,
    SCK: OutputPin,
    CS: OutputPin,
{
    /// Create a new BitBangSpi instance
    pub fn new(
        mosi: MOSI,
        miso: MISO,
        sck: SCK,
        cs: CS,
        config: SpiConfig,
    ) -> Result<Self, SpiError<MOSI::Error, MISO::Error, SCK::Error, CS::Error>> {
        let mut spi = Self {
            mosi,
            miso,
            sck,
            cs,
            config,
        };
        
        spi.initialize()?;
        Ok(spi)
    }
    
    /// Initialize the SPI pins
    fn initialize(&mut self) -> Result<(), SpiError<MOSI::Error, MISO::Error, SCK::Error, CS::Error>> {
        // Set initial states
        self.mosi.set_low().map_err(SpiError::Mosi)?;
        self.cs.set_high().map_err(SpiError::Cs)?;  // CS is active low
        
        // Set clock idle state based on CPOL
        if self.config.mode.cpol() {
            self.sck.set_high().map_err(SpiError::Sck)?;
        } else {
            self.sck.set_low().map_err(SpiError::Sck)?;
        }
        
        Ok(())
    }
    
    /// Transfer a single byte
    pub fn transfer_byte(&mut self, data_out: u8) -> Result<u8, SpiError<MOSI::Error, MISO::Error, SCK::Error, CS::Error>> {
        let mut data_in = 0u8;
        
        for i in 0..8 {
            let bit_index = if self.config.msb_first { 7 - i } else { i };
            
            // Set MOSI based on bit value
            if (data_out >> bit_index) & 1 == 1 {
                self.mosi.set_high().map_err(SpiError::Mosi)?;
            } else {
                self.mosi.set_low().map_err(SpiError::Mosi)?;
            }
            
            // Clock cycle based on CPHA
            if !self.config.mode.cpha() {
                // CPHA=0: Sample on first edge, setup on second
                self.delay();
                self.clock_first_edge()?;
                
                // Read MISO
                if self.miso.is_high().map_err(SpiError::Miso)? {
                    data_in |= 1 << bit_index;
                }
                
                self.delay();
                self.clock_second_edge()?;
            } else {
                // CPHA=1: Setup on first edge, sample on second
                self.delay();
                self.clock_first_edge()?;
                self.delay();
                self.clock_second_edge()?;
                
                // Read MISO
                if self.miso.is_high().map_err(SpiError::Miso)? {
                    data_in |= 1 << bit_index;
                }
            }
        }
        
        Ok(data_in)
    }
    
    /// Transfer multiple bytes
    pub fn transfer(&mut self, data: &mut [u8]) -> Result<(), SpiError<MOSI::Error, MISO::Error, SCK::Error, CS::Error>> {
        // Assert CS (active low)
        self.cs.set_low().map_err(SpiError::Cs)?;
        thread::sleep(Duration::from_micros(1));
        
        // Transfer each byte
        for byte in data.iter_mut() {
            *byte = self.transfer_byte(*byte)?;
        }
        
        // Deassert CS
        thread::sleep(Duration::from_micros(1));
        self.cs.set_high().map_err(SpiError::Cs)?;
        
        Ok(())
    }
    
    /// Write data (ignore received data)
    pub fn write(&mut self, data: &[u8]) -> Result<(), SpiError<MOSI::Error, MISO::Error, SCK::Error, CS::Error>> {
        self.cs.set_low().map_err(SpiError::Cs)?;
        thread::sleep(Duration::from_micros(1));
        
        for &byte in data {
            self.transfer_byte(byte)?;
        }
        
        thread::sleep(Duration::from_micros(1));
        self.cs.set_high().map_err(SpiError::Cs)?;
        
        Ok(())
    }
    
    /// Read data (send dummy bytes)
    pub fn read(&mut self, buffer: &mut [u8]) -> Result<(), SpiError<MOSI::Error, MISO::Error, SCK::Error, CS::Error>> {
        self.cs.set_low().map_err(SpiError::Cs)?;
        thread::sleep(Duration::from_micros(1));
        
        for byte in buffer.iter_mut() {
            *byte = self.transfer_byte(0xFF)?;
        }
        
        thread::sleep(Duration::from_micros(1));
        self.cs.set_high().map_err(SpiError::Cs)?;
        
        Ok(())
    }
    
    // Private helper methods
    fn clock_first_edge(&mut self) -> Result<(), SpiError<MOSI::Error, MISO::Error, SCK::Error, CS::Error>> {
        if self.config.mode.cpol() {
            self.sck.set_low().map_err(SpiError::Sck)
        } else {
            self.sck.set_high().map_err(SpiError::Sck)
        }
    }
    
    fn clock_second_edge(&mut self) -> Result<(), SpiError<MOSI::Error, MISO::Error, SCK::Error, CS::Error>> {
        if self.config.mode.cpol() {
            self.sck.set_high().map_err(SpiError::Sck)
        } else {
            self.sck.set_low().map_err(SpiError::Sck)
        }
    }
    
    fn delay(&self) {
        thread::sleep(Duration::from_micros(self.config.delay_us));
    }
}

// Error type
#[derive(Debug)]
pub enum SpiError<MOSI_E, MISO_E, SCK_E, CS_E> {
    Mosi(MOSI_E),
    Miso(MISO_E),
    Sck(SCK_E),
    Cs(CS_E),
}

// Example SPI Device implementation
pub struct SpiDevice<SPI> {
    spi: SPI,
}

impl<MOSI, MISO, SCK, CS> SpiDevice<BitBangSpi<MOSI, MISO, SCK, CS>>
where
    MOSI: OutputPin,
    MISO: InputPin,
    SCK: OutputPin,
    CS: OutputPin,
{
    pub fn new(spi: BitBangSpi<MOSI, MISO, SCK, CS>) -> Self {
        Self { spi }
    }
    
    /// Read a register from the device
    pub fn read_register(&mut self, reg_addr: u8) -> Result<u8, SpiError<MOSI::Error, MISO::Error, SCK::Error, CS::Error>> {
        let mut buffer = [reg_addr | 0x80, 0x00];  // Read bit set, dummy byte
        self.spi.transfer(&mut buffer)?;
        Ok(buffer[1])
    }
    
    /// Write to a register
    pub fn write_register(&mut self, reg_addr: u8, value: u8) -> Result<(), SpiError<MOSI::Error, MISO::Error, SCK::Error, CS::Error>> {
        let buffer = [reg_addr & 0x7F, value];  // Write bit clear
        self.spi.write(&buffer)
    }
    
    /// Read multiple registers
    pub fn read_multiple_registers(&mut self, start_addr: u8, buffer: &mut [u8]) -> Result<(), SpiError<MOSI::Error, MISO::Error, SCK::Error, CS::Error>> {
        let mut cmd_buffer = vec![start_addr | 0x80];
        cmd_buffer.extend(vec![0u8; buffer.len()]);
        
        self.spi.transfer(&mut cmd_buffer)?;
        buffer.copy_from_slice(&cmd_buffer[1..]);
        
        Ok(())
    }
}

// Mock GPIO implementation for testing
#[derive(Debug)]
struct MockError;

struct MockPin {
    state: bool,
}

impl MockPin {
    fn new() -> Self {
        Self { state: false }
    }
}

impl OutputPin for MockPin {
    type Error = MockError;
    
    fn set_high(&mut self) -> Result<(), Self::Error> {
        self.state = true;
        Ok(())
    }
    
    fn set_low(&mut self) -> Result<(), Self::Error> {
        self.state = false;
        Ok(())
    }
}

impl InputPin for MockPin {
    type Error = MockError;
    
    fn is_high(&self) -> Result<bool, Self::Error> {
        Ok(self.state)
    }
    
    fn is_low(&self) -> Result<bool, Self::Error> {
        Ok(!self.state)
    }
}

// Example usage
fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mosi = MockPin::new();
    let miso = MockPin::new();
    let sck = MockPin::new();
    let cs = MockPin::new();
    
    let config = SpiConfig {
        mode: SpiMode::Mode0,
        delay_us: 1,
        msb_first: true,
    };
    
    let spi = BitBangSpi::new(mosi, miso, sck, cs, config)
        .map_err(|_| "Failed to initialize SPI")?;
    
    let mut device = SpiDevice::new(spi);
    
    // Write to register
    device.write_register(0x1E, 0xA5)
        .map_err(|_| "Failed to write register")?;
    
    // Read from register
    let value = device.read_register(0x0F)
        .map_err(|_| "Failed to read register")?;
    
    println!("Register value: 0x{:02X}", value);
    
    // Read multiple registers
    let mut buffer = [0u8; 6];
    device.read_multiple_registers(0x00, &mut buffer)
        .map_err(|_| "Failed to read multiple registers")?;
    
    println!("Multiple registers: {:02X?}", buffer);
    
    Ok(())
}